### PR TITLE
:sparkles: Implement TWAP Oracle

### LINKED ISSUE #11: Implement TWAP oracle for AAVE
Gov token = AAVE.

Deposit token = DAI.

### PR DESCRIPTION
**Pull request by acemasterjb**
## Summary
Closes #11

## Details
* Implements the Uniswap V3 Oracle as per [Uniswap's specs](https://docs.uniswap.org/protocol/concepts/V3-overview/oracle).
* Tests were written with AAVE/DAI used as an example
    * tests utilises Foundry's [forking cheatcodes](https://book.getfoundry.sh/cheatcodes/forking) to fork Eth Mainnet
    * requires an Eth RPC node URL such as from [Infura](http://infura.io/)
* README was updated to explain the RPC node URL requirement

### CODE REVIEW

File: src/libraries/math/NewFullMath.sol [hunk 1]
```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: UNLICENCED
```
Comment (lajarre) [inline src/libraries/math/NewFullMath.sol]: Where is this file coming from? Is it not possible to install it as a forge-install dependency?

Comment (acemasterjb) [reply src/libraries/math/NewFullMath.sol]: It's coming from https://github.com/Uniswap/v3-core where some contracts and libs have pragmas which are not compatible with molten's, including:

https://github.com/Uniswap/v3-core/blob/main/contracts/libraries/FullMath.sol
https://github.com/Uniswap/v3-core/blob/main/contracts/libraries/TickMath.sol

File: src/libraries/oracle/OracleLibrary.sol [hunk 2]
```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: UNLICENCED
```
Comment (lajarre) [inline src/libraries/oracle/OracleLibrary.sol:1]: Here as well let's write where it comes from.

Also, let's just move all the libraries in `libraries/uniswap-v3-oracle` as these libraries are specific to the uniswap codebase.

File: test/Oracle.t.sol [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+contract OracleLibraryTest is Test {
+    uint256 public mainnetFork;
+
+    function setUp() public {
+        string memory mainnetRPCUrl = vm.envString("MAINNET_RPC_URL");
+        mainnetFork = vm.createFork(mainnetRPCUrl, 15895873);
+    }
+
+    function testConsultWithForkTicks() public {
```
Comment (lajarre) [inline test/Oracle.t.sol:19]: Why this test? Looks useless.

File: test/Oracle.t.sol [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            OracleConsulter.PriceConsultancyParams(
+                forkedAaveWETHPool,
+                14400,
+                uint128(wethOutput),
+                forkedWETH,
+                forkedAave
+            )
+        );
+
+        assertEq(aaveOutput, 10875160537248185); // ~0.0109 AAVE/DAI
```
Comment (lajarre) [inline test/Oracle.t.sol:69]: Isn't this number going to change over time? In this case, let's just verify that `OracleConsulter.consultPriceAtTick` properly returns.

File: README.md [hunk 5]
```diff
@@ -75,11 +74,20 @@ Basically, **Buyers can redeem their mTokens for governance tokens once the lock
...
+
+**Note**: to test the oracle, you will need a **Ethereum Mainnet RPC** node address to run the forked network tests. Such as from [Infura](http://infura.io/).
+
+Once you have a mainnet RPC node url, set it to the `MAINNET_RPC_URL` variable in the `.env.example` file and rename the file to `.env`.
+
+If you would like to **omit the oracle tests** then instead of `forge test` run
+
+```bash
+forge test --no-match-contract Oracle  # or --nmc for short
+```
```
Comment (lajarre) [inline README.md:93]: We should also add some details about deployment, as this is getting more complex
```suggestion
## Deploy

First deploy the OracleConsulter library:

`forge create src/libraries/oracle/OracleConsulter.sol:OracleConsulter`

Then use its address to deploy the main contract:

`forge create src/MoltenFunding.sol:MoltenFunding --libraries src/libraries/oracle/OracleConsulter.sol:OracleConsulter:<OracleConsulter contract address> ...`
```

**Review by lajarre — APPROVED** [review]
Review comments are non-critical. Can be merged as-is.

File: test/Oracle.t.sol (hunk 4 above)
Comment (acemasterjb) [reply test/Oracle.t.sol:69]: The test is specific to a block number (see `setUp`) so the result in this test will always be the same.

File: README.md (hunk 5 above)
Comment (acemasterjb) [reply README.md:93]: As the oracle logic in this codebase are all libraries, they are not to be deployed.

Comment (acemasterjb) [reply README.md:93]: ...which is why functions which just perform calculations (`view` and `pure`) should be considered to be grouped into libs to reduce deployed contract code size.

**Comment by acemasterjb** [comment]
Closing this PR, discussions re. the oracle should be directed to https://github.com/butterymoney/molten-oracle.

**Pull request closed without merging**